[AArch64][SVE] Improve fixed-length addressing modes.

rj-jesus opened the pull request:
This is based on the discussion in https://github.com/llvm/llvm-project/pull/127500#issuecomment-2688633766 and subsequent comments.

When compiling VLS SVE, the compiler often replaces VL-based offsets with immediate-based ones. This leads to a mismatch in the allowed addressing modes due to SVE loads/stores generally expecting immediate offsets relative to VL. For example, given:
```c

svfloat64_t foo(const double *x) {
  svbool_t pg = svptrue_b64();
  return svld1_f64(pg, x+svcntd());
}
```

When compiled with `-msve-vector-bits=128`, we currently generate:
```gas
foo:
        ptrue   p0.d
        mov     x8, #2
        ld1d    { z0.d }, p0/z, [x0, x8, lsl #3]
        ret
```

Instead, we could be generating:
```gas
foo:
        ldr     z0, [x0, #1, mul vl]
        ret
```

Likewise for other types, stores, and other VLS lengths.

This patch attempts to achieve the above by extending `SelectAddrModeIndexedSVE` to let constants through when `vscale` is known.

rj-jesus committed:
Precommit tests

rj-jesus committed:
[AArch64][SVE] Improve VLS imm addressing modes.

When compiling VLS SVE, the compiler often replaces VL-based offsets
with immediate-based ones. This leads to a mismatch during isel since
SVE loads/stores generally expect immediate offsets relative to VL.

For example, given:
```c

svfloat64_t foo(const double *x) {
  svbool_t pg = svptrue_b64();
  return svld1_f64(pg, x+svcntd());
}
```

When compiled with `-msve-vector-bits=128`, we currently generate:
```gas
foo:
        ptrue   p0.d
        mov     x8, #2
        ld1d    { z0.d }, p0/z, [x0, x8, lsl #3]
        ret
```

In practice, we could instead be generating:
```gas
foo:
        ldr     z0, [x0, #1, mul vl]
        ret
```

Likewise for other types, stores, and other VLS lengths.

paulwalker-arm commented:
Very nice.  Does this mean there's new hope for teaching `AArch64LoadStoreOpt` to convert contiguous SVE fill/spill instructions to NEON ldp/stp?

rj-jesus commented:
> Very nice. Does this mean there's new hope for teaching `AArch64LoadStoreOpt` to convert contiguous SVE fill/spill instructions to NEON ldp/stp?

Yep, that was also my thinking. I think it should work now.

david-arm commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -7379,13 +7379,27 @@ bool AArch64DAGToDAGISel::SelectAddrModeIndexedSVE(SDNode *Root, SDValue N,
-  SDValue VScale = N.getOperand(1);
-  if (VScale.getOpcode() != ISD::VSCALE)
+  int64_t MulImm = std::numeric_limits<int64_t>::max();
+  if (SDValue VScale = N.getOperand(1); VScale.getOpcode() == ISD::VSCALE)
```
I'd prefer the declaration of `VScale` to be moved out above the if statement for two reasons:

1. I personally find it hard to parse with the semi-colon because even outside of 'if' statements we don't normally have multiple statements on a single line.
2. You can reuse `VScale` in the else if case below, i.e. `auto C = dyn_cast<ConstantSDNode>(VScale)`

rj-jesus replied:
Thank you for the suggestion, done!

david-arm commented on llvm/test/CodeGen/AArch64/sve-fixed-length-offsets.ll:
```diff
@@ -0,0 +1,187 @@
+; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sve -aarch64-sve-vector-bits-min=1024 -aarch64-sve-vector-bits-max=1024 < %s | FileCheck %s --check-prefix=CHECK-1024
+; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sve -aarch64-sve-vector-bits-min=2048 -aarch64-sve-vector-bits-max=2048 < %s | FileCheck %s --check-prefix=CHECK-2048
+
+define void @nxv16i8(ptr %ldptr, ptr %stptr) {
```
Is it worth adding a test for an unpacked type too, i.e. loads such as <vscale x 4 x i8>, <vscale x 2 x float>? If I understand correctly it shouldn't end up using ldr/str because the memory type is too small.

Also, perhaps worth a test demonstrating legalisation, such as loads of <vscale x 4 x double>. I'm not sure in this case if we'd still lower to multiple ldrs?

rj-jesus committed:
Add unpacked/overpacked tests and move up VScale

madhur13490 commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -7380,12 +7380,27 @@ bool AArch64DAGToDAGISel::SelectAddrModeIndexedSVE(SDNode *Root, SDValue N,
   SDValue VScale = N.getOperand(1);
-  if (VScale.getOpcode() != ISD::VSCALE)
+  int64_t MulImm = std::numeric_limits<int64_t>::max();
+  if (VScale.getOpcode() == ISD::VSCALE)
```
This should be enclosed in `{}` to keep it unified with the `else` block per coding standards [guidelines](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements).

rj-jesus committed:
Add braces around if

rj-jesus committed:
Change MulImm to std::optional<int64_t> (NFC)

I think making `MulImm` std::optional helps make the intent of the
if-else chain clearer.

If anyone doesn't agree please let me know and I'll undo this.

david-arm commented on llvm/test/CodeGen/AArch64/sve-fixed-length-offsets.ll:
```diff
@@ -0,0 +1,362 @@
+; CHECK-128:       // %bb.0:
+; CHECK-128-NEXT:    add x8, x0, #128
+; CHECK-128-NEXT:    ldr z1, [x0, #8, mul vl]
+; CHECK-128-NEXT:    ldr z0, [x8, #1, mul vl]
```
So this looks correct, although a shame it's not just:

```
; CHECK-128-NEXT:    ldr z1, [x0, #8, mul vl]
; CHECK-128-NEXT:    ldr z0, [x0, #9, mul vl]
```

However, it's probably still better than before. Perhaps something we can improve later?

paulwalker-arm replied:
Personally I find this use of optional unnecessarily verbose.

rj-jesus replied:
Sounds good, I'll undo it, thanks!

david-arm reviewed (approved):
LGTM!

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -7380,17 +7380,31 @@ bool AArch64DAGToDAGISel::SelectAddrModeIndexedSVE(SDNode *Root, SDValue N,
+  } else if (auto C = dyn_cast<ConstantSDNode>(VScale)) {
+    int64_t ByteOffset = C->getSExtValue();
+    constexpr auto SVEBitsPerBlock = AArch64::SVEBitsPerBlock;
+    auto MinVScale = Subtarget->getMinSVEVectorSizeInBits() / SVEBitsPerBlock;
+    auto MaxVScale = Subtarget->getMaxSVEVectorSizeInBits() / SVEBitsPerBlock;
+
+    if (!MaxVScale || MinVScale != MaxVScale || ByteOffset % MaxVScale != 0)
```
This might start being a common idiom.  Do you mind adding a helper function to AArch64Subtarget along the lines of `optional<unsigned> getSVEVectorSizeInBits()`?

rj-jesus replied:
Thanks, that's a good idea. Would you prefer `optional`, or should we follow the logic that `getMinSVEVectorSizeInBits` and `getMaxSVEVectorSizeInBits` already use and return 0 when nothing is known about the limit?

paulwalker-arm replied:
I'm happy either way. Feel free to choose the option (pun intended :) ) you prefer.

rj-jesus replied:
Thanks - I've added this. Please let me know if that's what you had in mind. :)

rj-jesus committed:
Revert "Change MulImm to std::optional<int64_t> (NFC)"

This reverts commit 07d55840eeedeb5741b7cb5210d2c271a5060da5.

rj-jesus committed:
Add getSVEVectorSizeInBits to query known SVE length

paulwalker-arm reviewed (approved):
Other than a potentially erroneous assert this looks good to me.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64Subtarget.h:
```diff
@@ -405,6 +405,17 @@ class AArch64Subtarget final : public AArch64GenSubtargetInfo {
+  unsigned getSVEVectorSizeInBits() const {
+    assert(isSVEorStreamingSVEAvailable() &&
+           "Tried to get SVE vector length without SVE support!");
+    if (MaxSVEVectorSizeInBits &&
+        MinSVEVectorSizeInBits == MaxSVEVectorSizeInBits)
```
Thanks very much for the suggestion; that looks much better.
Should we let through the case `!MinSVEVectorSizeInBits && MaxSVEVectorSizeInBits == 128` too?

paulwalker-arm replied:
Not sure if this can happen outside of handwritten IR but otherwise it sounds like a reasonable suggestion to me.

rj-jesus replied:
I've left it as is for now due to the lack of a motivating example and to keep it consistent with `getMinSVEVectorSizeInBits`/`getMinSVEVectorSizeInBits`, which I suppose could return 128/2048 as the architecture bounds and avoid this problem altogether. Please let me know if you'd rather I add it.

I'll let the tests run and assuming they are OK merge it.

rj-jesus committed:
Remove assert and redundant check

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -7380,12 +7380,26 @@ bool AArch64DAGToDAGISel::SelectAddrModeIndexedSVE(SDNode *Root, SDValue N,
 
+  assert(MulImm != std::numeric_limits<int64_t>::max() &&
+         "Uninitialized MulImm.");
+
```
Sounds good to me, removed.

omjavaid commented:
This PR appears to have broken [clang-aarch64-sve-vls](https://lab.llvm.org/buildbot/#/builders/143) buildbot. Here is the failing build https://lab.llvm.org/buildbot/#/builders/143/builds/5952
